LCORE-38: document BYOK (#92380)
openshift/openshift-docs: modules/ols-providing-custom-knowledge-to-the-llm.adoc

// Module included in the following assemblies:
// * lightspeed-docs-main/configure/ols-configuring-openshift-lightspeed.adoc

:_mod-docs-content-type: PROCEDURE
[id="providing-custom-knowledge-to-the-llm_{context}"]
= Providing custom knowledge to the LLM

Customize the information available to the large language model (LLM) by providing access to a container image that resides in a remote image registry. The examples in this procedure use `quay.io` as the remote container image registry, and the path for the custom image is `quay.io/<username>/my-byok-image:latest`.

:FeatureName: The BYO Knowledge feature

> include::snippets/technology-preview.adoc[]

Reviewer: Should this go at the level of the concept? I am actually not sure if have an opinion on this as an org, but seems it might be good to know on first appearance. Although making sure they know before trying the steps also seems reasonable.

Author: @jeana-redhat I don't think we have an established norm for this. My thought was it might be more impactful to call this out with the instructions someone will use. I'm not sure people will always read the concept and then the task. If we call this out in the task, I think there is a better chance customers will see this when they perform the action.

I have another TP feature in the doc set up this way. If I change it here I should change that one too.

Reviewer: yeah I am fine with it either way - keeping it close to the action works for me :)

.Prerequisites

> * You are logged in to the {ocp-product-title} web console as a user with the `cluster-admin` role. Alternatively, you are logged in to a user account that has permission to create a cluster-scoped CR file.

Reviewer: idk about providing alternatives here per se. To me, I think you can just say:
```suggestion
* You are logged in to the {ocp-product-title} web console as a user account that has permission to create a cluster-scoped custom resource (CR) file.
```
or, if it's important/helpful to point out that the cluster admin has those perms, something like this?
```suggestion
* You are logged in to the {ocp-product-title} web console as a user account that has permission to create a cluster-scoped custom resource (CR) file, such as a user with the `cluster-admin` role.
```

* You have installed the {ols-long} Operator.

> * You have configured the large language model (LLM) provider.

Reviewer: already expanded above
```suggestion
* You have configured the LLM provider.
```

Author: @jeana-redhat  I'm all for making this change - I think LLM is widely understood by users in the AI space. I thought we were supposed to write out first instance in each topic. Just wanted to check with you and see if that is not the case.

> * The custom information you want to add resides as a collection of Markdown files; no other file format is supported.

Reviewer: (localization prefers no `;`)
```suggestion
* The custom information you want to add resides as a collection of Markdown files. No other file format is supported.
```

Author: Thanks, I did not know that. I have not been part of localizing with this organization and have not heard that at other organizations.

Done.

* You have an account for a container image registry, such as `quay.io`.

.Procedure 

> . Specify the location of the directory with the markdown files for the custom RAG database and the path for the image that the BYO Knowledge tool generates by running the following command: 

Reviewer: ```suggestion
. Specify the location of the directory with the Markdown files for the custom retrieval-augmented generation (RAG) database and the path for the image that the BYO Knowledge tool generates by running the following command: 
```

Author: done

+
[source,terminal]
----

> $ podman run -e OUT_IMAGE_TAG=my-byok-image -it --rm --device=/dev/fuse -v <directory_for_markdown_files>:/markdown:Z -v <directory_for_image_tar>:/output:Z $MY_BYOK_TOOL_IMAGE

Reviewer: Consider splitting the arguments across lines for scannability i.e.
```suggestion
$ podman run \
  -e OUT_IMAGE_TAG=my-byok-image \
  -it --rm --device=/dev/fuse \
  -v <directory_for_markdown_files>:/markdown:Z \
  -v <directory_for_image_tar>:/output:Z $MY_BYOK_TOOL_IMAGE
```
In doing this, I notice it's got two `-v`s - is that right?

Author: Yes, ran it by the dev to double-check, too.

done.

----
+
[NOTE]
====
The container image that the tool generates contains a custom RAG database. The RAG database provides additional information to the LLM. 
====

. Load the container image that the BYO Knowledge tool generated by running the following command:
+
[source,terminal]
----
$ podman load < <directory_for_image_tar>/<my-byok-image.tar>
----

> . Display the podman images that reside on your laptop by running the following command:

Reviewer: ```suggestion
. Display the podman images that are on your local computer by running the following command:
```
or maybe "exist"? idk, "reside" seems a bit anthropomorphic

Author: done

+
[source,terminal]
----
$ podman images
----
+
.Example output
[source,terminal]
----
REPOSITORY                            TAG                IMAGE ID      CREATED       SIZE
localhost/my-byok-image               latest             be7d1770bf10  1 minute  ago    2.37 GB
...
----

> . Tag the local image with a new name and destination so that the image can be pushed to the container image registry by running the following command:

Reviewer: typically we avoid "new" as a fluff word. I can't tell if you need it here or not, if you need it, keep it :)
```suggestion
. Tag the local image with a name and destination so that the image can be pushed to the container image registry by running the following command:
```

Author: done

+
[source,terminal]
----
$ podman tag localhost/my-byok-image:latest quay.io/<username>/my-byok-image:latest
----

. Push the local container image to the container image registry by running the following command:
+
[source,terminal]
----
$ podman push quay.io/<username>/my-byok-image:latest
----

> . Modify the `OLSconfig` custom resource to deploy the newly created RAG database alongside the existing one:

Reviewer: ```suggestion
. Modify the `OLSconfig` CR to deploy the newly created RAG database alongside the existing one:
```

Author: done

.. In the {ocp-product-title} web console, click *Operators* -> *Installed Operators*. 

.. Select *All Projects* in the  *Project* dropdown at the top of the screen.

> .. Click {ols-long} Operator.

Reviewer: Should the title be bold as a UI item in this case?
```suggestion
.. Click *{ols-long} Operator*.
```

Author: done

.. Click *OLSConfig*, then click the `cluster` configuration instance in the list.

.. Click the *YAML* tab.

.. Insert the `spec.ols.rag` yaml code:
+
.Example `OLSconfig` CR file
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: ols.openshift.io/v1alpha1
kind: OLSConfig
metadata:
  name: cluster
spec:
  ols:
    rag:
      - image: quay.io/<username>/my-byok-image:latest <1>
        indexPath: /rag/vector_db
        indexID: vector_db_index    
----

> <1> Specifies the tag for the image that was pushed to the image registry so that the {ols-long} Operator can access the custom content. The {ols-long} Operator can work with more than one RAG database that you create.

Reviewer: Consider doing this as a "where" item instead of a callout

Author: done

. Click *Save*.

.Verification

> . Access the {ols-long} virtual assistant and submit a question associated with the custom content that was provided to the LLM.

Reviewer: ```suggestion
* Access the {ols-long} virtual assistant and submit a question associated with the custom content that you provided to the LLM.
```

Author: done
